addons/point_of_sale/static/src/app/services/pos_store.js

```diff
@@ -2715,6 +2720,275 @@ export class PosStore extends WithLazyGetterTrap {
         });
         await validation.validateOrder(false);
     }
+
+    handlePreparationHistory(srcPrep, destPrep, srcLine, destLine, qty) {
+        const srcKey = srcLine.preparationKey;
+        const destKey = destLine.preparationKey;
+        const srcQty = srcPrep[srcKey]?.quantity;
+
+        if (srcQty) {
+            if (srcQty <= qty) {
+                const newPrep = { ...srcPrep[srcKey], uuid: destLine.uuid };
+                destPrep[destKey] = newPrep;
+                delete srcPrep[srcKey];
+            } else {
+                srcPrep[srcKey].quantity = srcQty - qty;
+                destPrep[destKey] = { ...srcPrep[srcKey], uuid: destLine.uuid, quantity: qty };
+            }
+        }
+    }
+
+    get isSelectedLineCombo() {
+        const selectedOrderline = this.selectedOrder.getSelectedOrderline();
+        return selectedOrderline && selectedOrderline.isPartOfCombo();
+    }
+
+    get applicableProductCombo() {
+        const productInOrder = {};
+        this.selectedOrder.lines.forEach((l) => {
+            if (l.combo_parent_id || (l.combo_line_ids && l.combo_line_ids.length)) {
+                return;
+            }
+            const pid = l.product_id.id;
+            if (!productInOrder[pid]) {
+                productInOrder[pid] = {};
+            }
+            productInOrder[pid][l.uuid] = l.qty;
+        });
+        const matchingCombos = [];
+        for (const comboProduct of this.models["product.template"].filter(
+            (p) => p.type == "combo"
+        )) {
```

Comment:
This will cause performance issue with a lot of products.

You can maybe use `databaseIndex` of data service to precompute all values.